docs/proposals/21-ipv6-singlestack-local.md

```diff
@@ -0,0 +1,213 @@
+
+Also, gardener-scheduler chooses seeds with `ipFamilies` settings matching the shoot's configuration.
+
+#### DNS Records
+
+gardenlet creates `DNSRecord` objects for the shoot API server with the record type corresponding to the `Shoot.spec.networking.ipFamilies` setting (`A` for IPv4, `AAAA` for IPv6).
+`AAAA` records are created with a suffix to avoid name collisions in a potential dual-stack implementation.
+
+#### Network Policies
+
+Gardener manages several `NetworkPolicy` objects with static CIDRs (e.g. `allow-to-public-networks` in seeds).
+For these, IPv6-equivalents of the IPv4 CIDRs are added, e.g., `::/0` as an equivalent to `0.0.0.0/0`.
+
+#### Shoot Worker Node Kernel Configuration
+
+Kubernetes networking requires IPv6 forwarding to be enabled on the OS level.
+Hence, gardenlet explicitly enables the corresponding kernel setting for shoot worker nodes via `OperatingSystemConfigurations`, similar to IPv4 (ref [gardener/gardener#7046](https://github.com/gardener/gardener/pull/7046)).
+
+#### Docker Hub Images
+
+The `docker.io` registry doesn't support pulling images over IPv6 (see [Beta IPv6 Support on Docker Hub Registry](https://www.docker.com/blog/beta-ipv6-support-on-docker-hub-registry/)).
+
+Container images from `docker.io` used on shoots and seeds are rewritten to `registry.ipv6.docker.com` if the corresponding `ipFamilies` field specifies IPv6 singe-stack.
+
+#### E2E Tests
+
+Small changes in the OS and networking stack might lead to regressions of the IPv6 feature.
+Hence, [e2e tests](../development/testing.md#end-to-end-e2e-tests-using-provider-local) (presubmits and periodics) are added to prevent unnoticed regressions of the feature.
+
+#### Networking Extensions
+
+Networking extensions need to support the configuration of IPv6-related settings of the networking implementation in their `NetworkConfig` API.
+This is not specified any further by gardener and will be implemented differently in different extensions (as the `NetworkConfig` API design differs already today).
+In general, extensions need to respect the `Shoot.spec.networking.ipFamilies` settings in all aspects: API validation, defaulting, and corresponding handling in code (i.e., configuring IPAM, etc.).
+
+## Alternatives Considered
+
+Instead of adding IPv6-equivalent fields in the `Shoot` and `Seed` API (e.g., `Shoot.spec.networking.{pods,services,nodes}V6`), new arrays could be added that allow specifying both settings in a single list (similar to `Service.spec.clusterIPs`).
+However, to guarantee backward and forward compatibility, the new array would need to be kept in sync with the existing field and the existing field must not be removed.
+This makes API validation more complex and might confuse users.
+For the sake of simplicity and a more expressive and less confusing API, this approach was discarded.
+However, a potential `core.gardener.cloud/v1` API might choose to use a single array for both settings instead of separate fields.
```

Comment:
Technically it is possible to detect the addressfamily of the given cidrs {pods,service,nodes}. The validation logic must then check if all given cidrs are from the same addressfamily in the single stack case, and in the dual stack case if for {pods, services and nodes } cidrs for both addressfamilies are given.

This would allow us to keep the existing [shoot,seed].spec.networking as it is (execept that pods,services,nodes are slices now).

But i mostly agree, keeping it explicit from the user perspective is less errorprone and will lead to a simpler to understand implementation.

Reply:
You're right, this was also our reasoning.
The main thing, that kept us from re-using the existing fields, is that we would need to model a structure (list) in a string field (comma-separated) for the dual-stack case.
While this works, it doesn't fit the Kubernetes API design principles.
And as you say: it definitely makes the implementation simpler and less error-prone if you have separate fields instead of a single field that you need to parse and juggle around every time :)

Reply:
Then the question remains whether we should support IPv6 addresses in "old" the `pod/node/service` specs for the single-stack use case. Yes, I agree that for the dual-stack case, being more explicit about IPv4 and IPv6 and avoiding the trouble of a two-element list sounds legit, but having `pods` meaning `podsV4` is also somewhat inconsistent.
Do we need to consider migration paths here or can we assume the spec is only ever defined at cluster creation?

Reply:
> having `pods` meaning `podsV4` is also somewhat inconsistent.

I kind of agree. However, we can't change the field names for the sake of compatibility. Also, I think it's not very confusing, as the existing fields are well-known.

> Do we need to consider migration paths here or can we assume the spec is only ever defined at cluster creation?

We shouldn't consider migration paths right now as written in the motivation section.
However, it's confusing and somewhat incompatible if we now allow specifying IPv6 CIDRs in `{pods,nodes,services}` for IPv6 single-stack and for dual-stack IPv6 CIDRs must be specified in `{pods,nodes,services}V6`.

From our perspective, having dedicated fields and keeping semantics in all possible cases of single-stack and dual-stack will be the most straightforward, explicit, and least error-prone.
I know, this decision is the most controversial part of the GEP, so I'm happy for all the feedback!

Reply:
My preference would be to continue supporting IPv4 xor IPv6 in `{pods,nodes,services}` for compatibility the single-stack case and require using `{pods,nodes,services}V6` or `{pods,nodes,services}V4` in the dual-stack case (while recommending to use the explicit versions also for single-stack)

Reply:
> while recommending to use the explicit versions also for single-stack

This kind of split logic makes things more complicated. You will need to keep both fields in sync.
There shouldn't be two ways to configure the same thing. I would like to keep it explicit and straightforward as proposed.
I'm not sure what you're trying to improve here.